File: pset6/credit/credit.py
from cs50 import get_int

# Request input from user
num = str(get_int("Number: "))

singles = 0
doubles = 0
length = len(num)

if length != 15 and length != 16:
    print("INVALID")

else:
    # iterate along and sum doubles
    for i in range(length - 2, -1, -2):
        temp = str(int(num[i]) * 2)
        for y in range(len(temp)):
            doubles += int(temp[y])

    # interate along and sum singes
    for i in range(length - 1, -1, -2):
        singles += int(num[i])

    # Store first and first two digits
    first = num[0]
    second = num[0] + num[1]

    # Check if luhn value is is correct
    luhn = str(singles + doubles)
    if luhn[-1] != "0":
        print("INVALID")

    # Figure out what type of card
    elif length == 15 and (second == "34" or second == "37"):
        print("AMEX")

    elif len(num) == 16 and (second == "51" or second == "52" or second == "53" or second == "54" or second == "55"):
        print("MASTERCARD")

    elif (len(num) == 16 or len(num) == 13) and first == "4":
        print("VISA")

    else:
        print("INVALID")

File: pset1/credit/credit.c
#include <stdio.h>
#include <cs50.h>
#include <math.h>

int main(void)
{

    //Request user input - consider what checks are required
    long long cardno = get_long_long("Number: ");

    //Initialise working variables
    long long length = 0;
    long long sumsingle = 0;
    long long sumdub = 0;
    long long singles = cardno;
    long long doubles = cardno / 10;
    long long lennum = cardno;

    //Calculate sum of singles
    while (singles != 0)
    {

        //Sum alt digits starting with last
        sumsingle += singles % 10;
        //printf("digit = %lli\n", singles%10);
        singles /= 100;
        //printf("sumsingle = %lli\n", sumsingle);
    }
    //Calculate sum of doubles
    while (doubles != 0)
    {
        long long dubvalue = (doubles % 10) * 2;
        sumdub += dubvalue % 10 + (dubvalue / 10) % 10;
        //printf("digit = %lli\n", doubles%10);
        doubles /= 100;
        //printf("sumdouble = %lli\n", sumdub);
    }


    //Calulate length
    while (lennum != 0)
    {
        length ++;
        lennum /= 10;
        //printf("length = %lli\n", length);
    }

    //Determine 1st and 2nd digits
    long long first = cardno / pow(10, length - 1);
    long long second = cardno / pow(10, length - 2);
    //printf("1stnum = %lli\n", first);
    //printf("2ndnum = %lli\n", second);

    //Check if valid luhn
    if ((sumsingle + sumdub) % 10 == 0)
    {
        //Check if AMEX
        if (length == 15 && (second == 34 || second == 37))
        {
            printf("AMEX\n");
        }
        //Check if MASTERCARD
        else if (length == 16 && (second == 51 || second == 52 || second == 53 || second == 54 || second == 55))
        {
            printf("MASTERCARD\n");
        }
        //Check if VISA
        else if ((length == 16 || length == 13) && first == 4)
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }




}